src/mscorlib/shared/System/IO/PathHelper.Windows.cs

```diff
@@ -24,63 +25,73 @@ internal class PathHelper
-                    if (fullPath.Length == path.Length && fullPath.StartsWith(path))
+                    if (builder.AsSpan().Equals(path.AsReadOnlySpan()))
                     {
                         // If we have the exact same string we were passed in, don't bother to allocate another string from the StringBuffer.
                         return path;
                     }
-                    return fullPath.ToString();
+                    return builder.ToString();
                 }
             }
             finally
             {
                 // Clear the buffer
-                fullPath.Free();
+                builder.Dispose();
             }
         }
 
-        private static unsafe void GetFullPathName(string path, ref StringBuffer fullPath)
+        private static void GetFullPathName(string path, ref ValueStringBuilder builder)
         {
             // If the string starts with an extended prefix we would need to remove it from the path before we call GetFullPathName as
             // it doesn't root extended paths correctly. We don't currently resolve extended paths, so we'll just assert here.
             Debug.Assert(PathInternal.IsPartiallyQualified(path) || !PathInternal.IsExtended(path));
+            builder.EnsureCapacity(path.Length);
+            builder.Append('\0');
 
-            fixed (char* pathStart = path)
+            uint result = 0;
+            while ((result = Interop.Kernel32.GetFullPathNameW(path, (uint)builder.Capacity, ref builder[0], IntPtr.Zero)) > builder.Capacity)
             {
-                uint result = 0;
-                while ((result = Interop.Kernel32.GetFullPathNameW(pathStart, (uint)fullPath.Capacity, fullPath.UnderlyingArray, IntPtr.Zero)) > fullPath.Capacity)
-                {
-                    // Reported size is greater than the buffer size. Increase the capacity.
-                    fullPath.EnsureCapacity(checked((int)result));
-                }
-
-                if (result == 0)
-                {
-                    // Failure, get the error and throw
-                    int errorCode = Marshal.GetLastWin32Error();
-                    if (errorCode == 0)
-                        errorCode = Interop.Errors.ERROR_BAD_PATHNAME;
-                    throw Win32Marshal.GetExceptionForWin32Error(errorCode, path);
-                }
+                // Reported size is greater than the buffer size. Increase the capacity.
+                builder.EnsureCapacity(checked((int)result));
+            }
 
-                fullPath.Length = checked((int)result);
+            if (result == 0)
+            {
+                // Failure, get the error and throw
+                int errorCode = Marshal.GetLastWin32Error();
+                if (errorCode == 0)
+                    errorCode = Interop.Errors.ERROR_BAD_PATHNAME;
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, path);
             }
+
+            builder.Length = checked((int)result);
```

Comment:
Nit: This does not need to be checked.

Reply:
Because per spec we can't get an overflow value? (Should never ask for more than UNICODE_STRING can hold). I was thinking technicalities here as the API returns uint by definition.

Reply:
The while loop above will exit only if the total size is less than the capacity. So it has to fit into int.

Reply:
:) thanks!